### PR TITLE
Preservation fix

### PR DESCRIPTION
**Pull request by mrvoorhe**
If duplicate preservations, keep the one that results in preserving the most

### CODE REVIEW

File: linker/Mono.Linker/Annotations.cs [hunk 1]
```diff
@@ -139,7 +139,31 @@ public bool IsPreserved (TypeDefinition type)
...
 		public void SetPreserve (TypeDefinition type, TypePreserve preserve)
 		{
-			preserved_types [type] = preserve;
+			if (preserved_types.ContainsKey (type))
+				preserved_types [type] = ChoosePreserveActionWhichPreservesTheMost (preserved_types [type], preserve);
+			else
+				preserved_types [type] = preserve;
+		}
+
+		public static TypePreserve ChoosePreserveActionWhichPreservesTheMost (TypePreserve leftPreserveAction, TypePreserve rightPreserveAction)
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: Why is it public?

File: linker/Mono.Linker/Annotations.cs [hunk 2]
```diff
@@ -139,7 +139,31 @@ public bool IsPreserved (TypeDefinition type)
...
+		{
+			if (leftPreserveAction == rightPreserveAction)
+				return leftPreserveAction;
+
+			if (leftPreserveAction == TypePreserve.All || rightPreserveAction == TypePreserve.All)
+				return TypePreserve.All;
+
+			if (leftPreserveAction == TypePreserve.Nothing && rightPreserveAction != TypePreserve.Nothing)
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: The && seems redundant

File: linker/Mono.Linker/Annotations.cs [hunk 3]
```diff
@@ -139,7 +139,31 @@ public bool IsPreserved (TypeDefinition type)
...
+				return rightPreserveAction;
+
+			if (rightPreserveAction == TypePreserve.Nothing && leftPreserveAction != TypePreserve.Nothing)
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: The && seems redundant

File: linker/Mono.Linker/Annotations.cs [hunk 4]
```diff
@@ -139,7 +139,31 @@ public bool IsPreserved (TypeDefinition type)
 
 		public void SetPreserve (TypeDefinition type, TypePreserve preserve)
 		{
-			preserved_types [type] = preserve;
+			if (preserved_types.ContainsKey (type))
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: Use `TryGetValue` instead

File: linker/Mono.Linker/Annotations.cs [hunk 5]
```diff
@@ -139,7 +139,31 @@ public bool IsPreserved (TypeDefinition type)
...
+				preserved_types [type] = ChoosePreserveActionWhichPreservesTheMost (preserved_types [type], preserve);
+			else
+				preserved_types [type] = preserve;
```
Comment (marek-safar) [inline linker/Mono.Linker/Annotations.cs]: This should be `Add`

File: linker/Mono.Linker/Annotations.cs (hunk 1 above)
Comment (mrvoorhe) [inline linker/Mono.Linker/Annotations.cs]: Unit tests.  I'll bring over the tests from unity into the new test framework

File: linker/Mono.Linker/Annotations.cs (hunk 2 above)
Comment (mrvoorhe) [inline linker/Mono.Linker/Annotations.cs]: Agreed.  I'll remove them

**Comment by mrvoorhe** [comment]
@marek-safar The PR has been updated with the changes you requested

**Pull request merged**